reana_workflow_engine_serial/tasks.py

```diff
@@ -66,151 +61,169 @@ def load_json(ctx, param, value):
-                                               "job_ids": []},
-                                          }})
-
-    expanded_workflow_json = serial_load(None, workflow_json,
+    return workflow_workspace, publisher, cache_enabled
+
+
+def run(workflow_json,
+        workflow_parameters,
+        operational_options,
+        workflow_uuid,
+        workflow_workspace,
+        publisher,
+        cache_enabled):
+    """Run a serial workflow."""
+    expanded_workflow_json = serial_load(None,
+                                         workflow_json,
                                          workflow_parameters)
 
-    current_command_idx = 0
+    publish_workflow_start(workflow_json,
+                           workflow_uuid,
+                           publisher)
+
     for step_number, step in enumerate(expanded_workflow_json['steps']):
-        last_command = 'START'
-        for command in step['commands']:
-            current_command_idx += 1
-            job_spec = {
-                "experiment": os.getenv("REANA_WORKFLOW_ENGINE_EXPERIMENT",
-                                        "default"),
-                "image": step["environment"],
-                "cmd": "bash -c \"cd {0} ; {1} \"".format(
-                    workflow_workspace, escape_shell_arg(command)),
-                "prettified_cmd": command,
-                "workflow_workspace": workflow_workspace,
-                "job_name": command,
-                "cvmfs_mounts": []
-            }
-            if MOUNT_CVMFS:
-                job_spec["cvmfs_mounts"] = ["cms", "atlas", "alice", "lhcb"]
-            job_spec_copy = dict(job_spec)
-            clean_cmd = ';'.join(job_spec_copy['cmd'].split(';')[1:])
-            job_spec_copy['cmd'] = clean_cmd
-            if 'CACHE' not in operational_options or \
-                    operational_options.get('CACHE').lower() != 'off':
-                http_response = rjc_api_client.check_if_cached(
-                    job_spec_copy,
-                    step,
+        run_step(step_number,
+                 step,
+                 workflow_workspace,
+                 cache_enabled,
+                 expanded_workflow_json,
+                 workflow_json,
+                 publisher,
+                 workflow_uuid)
+
+
+def run_step(step_number,
+             step,
+             workflow_workspace,
+             cache_enabled,
+             expanded_workflow_json,
+             workflow_json,
+             publisher,
+             workflow_uuid):
+    """Run a step of a serial workflow."""
+    for command in step['commands']:
+        job_spec = build_job_spec(step['environment'],
+                                  command,
+                                  workflow_workspace,
+                                  MOUNT_CVMFS)
+        job_spec_copy = dict(job_spec)
+        job_spec_copy['cmd'] = sanitize_command(job_spec_copy['cmd'])
+
+        if cache_enabled:
+            cached_info = check_cache(rjc_api_client,
```

Comment:
When there is no cache hit `check_cache` is returning `False` so next if will fail. We should probably return an empty dict and in the next if do `if cached_info.get('result_path')`.

Reply:
Right, thanks :+1: !